{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b930953e-dbd7-4291-ae49-be3422ae9ead",
   "metadata": {
    "id": "b930953e-dbd7-4291-ae49-be3422ae9ead"
   },
   "source": [
    "## <u>__Exercises__</u> ##\n",
    "\n",
    "**Part 1**\n",
    "\n",
    "We can generate (pseudo-)random numbers using the `random` function:\n",
    "\n",
    "https://docs.python.org/3/library/random.html\n",
    "\n",
    "Can you combine this with a loop and and `if` function to roll the dice (pick a random number between 1-6) 10 times in a row?\n",
    "Next, can you keep rolling a dice until you get a 6? How about two 6's in a row?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bd09fac-5ffb-485f-b446-ae31511b62ad",
   "metadata": {
    "id": "7bd09fac-5ffb-485f-b446-ae31511b62ad"
   },
   "outputs": [],
   "source": [
    "# @title Code snippet\n",
    "\n",
    "\n",
    "import random\n",
    "print(random.random())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14413e75-2bca-4549-a951-8899eff7b5cc",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "14413e75-2bca-4549-a951-8899eff7b5cc",
    "outputId": "13414e82-4a79-4b3b-f5ea-5ac127efe6d0"
   },
   "outputs": [],
   "source": [
    "# @title Solution - don't peak!\n",
    "\n",
    "import random\n",
    "\n",
    "x = 0\n",
    "print(\"Roll 10x: \")\n",
    "for x in range(10):\n",
    "    roll = random.randint(1, 6)\n",
    "    print(f\"Roll: {x+1}: Rolled: {roll}\")\n",
    "\n",
    "print()\n",
    "print(\"Roll a 6 and stop: \")\n",
    "rolls = 0\n",
    "while True:\n",
    "    roll = random.randint(1, 6)\n",
    "    rolls += 1\n",
    "    print(f\"Rolled: {roll}\")\n",
    "\n",
    "    if roll == 6:\n",
    "        print(f\"Got a 6 after {rolls} rolls!\")\n",
    "        break\n",
    "\n",
    "print()\n",
    "print(\"Roll until you get two 6's in a row: \")\n",
    "rolls = 0\n",
    "while True:\n",
    "    roll = random.randint(1, 6)\n",
    "    rolls += 1\n",
    "    print(f\"Rolled: {roll}\")\n",
    "\n",
    "    if roll == 6:\n",
    "        print(f\"Got a 6 after {rolls} rolls!\")\n",
    "        roll = random.randint(1, 6)\n",
    "        if roll == 6:\n",
    "            print(f\"Got a 6 the next roll!\")\n",
    "            break\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ff88f96-2ef7-4707-981b-ccabb5f7a3d1",
   "metadata": {
    "id": "6ff88f96-2ef7-4707-981b-ccabb5f7a3d1"
   },
   "source": [
    "**Part 2**\n",
    "\n",
    "If the random function is truely random, we should get equal numbers of 1-6 after suitably many dice rolls.\n",
    "\n",
    "Use an array or list to capture the output of the dice rolls and see how many rolls are needed before the number of rolls is (roughly) even.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8VG92l7Hk-Xl",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "8VG92l7Hk-Xl",
    "outputId": "f381f396-7dce-4c7f-a5ae-037f673e2afe"
   },
   "outputs": [],
   "source": [
    "# @title Code snippet\n",
    "# count may be useful here:\n",
    "\n",
    "list = [1,1,2,2,3,5]\n",
    "print(list.count(1))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53577bfb-1dbc-4a52-b95c-cdcd441746a7",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 490
    },
    "id": "53577bfb-1dbc-4a52-b95c-cdcd441746a7",
    "outputId": "74633f25-aee4-478e-a0a8-ec3be1c4e680"
   },
   "outputs": [],
   "source": [
    "# @title A solution\n",
    "\n",
    "import random\n",
    "import matplotlib.pyplot as plt # we will do more plotting tomorrow\n",
    "\n",
    "x = 0\n",
    "nrolls = 10000   # Number of dice rolls\n",
    "outcome = []  # Our list of roll outcomes. We start with an empty list.\n",
    "\n",
    "for x in range(nrolls):\n",
    "    roll = random.randint(1, 6)\n",
    "    outcome.append(roll)\n",
    "\n",
    "#Â print(outcome)   #check that we are adding dice rolls to the list\n",
    "\n",
    "counts = []\n",
    "for i in range(1,7):\n",
    "    counts.append(outcome.count(i))\n",
    "print(counts)\n",
    "\n",
    "\n",
    "plt.bar(range(1,7),counts)\n",
    "plt.title('A graph...')\n",
    "plt.xlabel('number')\n",
    "plt.ylabel('count')\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3oQJFDdhovFM",
   "metadata": {
    "id": "3oQJFDdhovFM"
   },
   "source": [
    "**Part 3**\n",
    "\n",
    "No let's do something a little more biological.\n",
    "Generate a random sequence of DNA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "MqUw76bKmkNK",
   "metadata": {
    "id": "MqUw76bKmkNK"
   },
   "outputs": [],
   "source": [
    "# @title A solution\n",
    "\n",
    "import random\n",
    "\n",
    "length = 100\n",
    "\n",
    "list = []\n",
    "gene = []\n",
    "\n",
    "# creat a list of numbers 1-4 of length 'length'\n",
    "for i in range(length):\n",
    "  list.append(random.randint(1, 4))\n",
    "# print(list)\n",
    "\n",
    "# replace 1-4 with A-G\n",
    "gene = list\n",
    "for i in range(len(gene)):\n",
    "  gene[i] = \"A\" if gene[i] == 1 else gene[i]\n",
    "  gene[i] = \"T\" if gene[i] == 2 else gene[i]\n",
    "  gene[i] = \"C\" if gene[i] == 3 else gene[i]\n",
    "  gene[i] = \"G\" if gene[i] == 4 else gene[i]\n",
    "\n",
    "# convert the list into a single string\n",
    "geneseq = \"\".join(gene)\n",
    "\n",
    "# check length using `len`\n",
    "print(\"gene of length: \",len(geneseq))\n",
    "print(geneseq)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "135e8487-c6de-46ac-9062-12517af5be01",
   "metadata": {},
   "source": [
    "**Part 4**\n",
    "\n",
    "Does the DNA sequence contain an ATG codon or Eco R1 restiction site (\"GAATTC\")?\n",
    "If not, can you make a loop that keeps generating new random sequences until you find one?\n",
    "\n",
    "For extra credit, can you calculate the melting temperature. \n",
    "\n",
    "https://www.calculator.bio/gc-content-tm/\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bed3ea05-30d6-49bf-8706-26e8fc839bb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Another way using random.choices()\n",
    "# from https://colab.research.google.com/github/Intertangler/bioinformatics_stockholm/blob/master/1b_biopython_and_string_practice.ipynb#scrollTo=fqs2bnmxn1_4\n",
    "\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "def generate_random_sequence(sequence_length, probability_distribution):\n",
    "    sequence = \"\"\n",
    "    for i in range(0,sequence_length):\n",
    "        sequence = sequence + random.choices([\"A\",\"C\",\"G\",\"T\"], weights=probability_distribution,k=1)[0]\n",
    "    return sequence\n",
    "\n",
    "length = 10\n",
    "dist = [0.25,0.25,0.25,0.25] # probability of finding A, C, G and T, respectively\n",
    "\n",
    "\n",
    "print(\"look for 'GAATTC':\")\n",
    "\n",
    "max = 100 # instead of using while, we can use a for loop with a maximum of `max' attempts\n",
    "j = 0\n",
    "k = 0\n",
    "\n",
    "while True or j < max :\n",
    "    j += 1\n",
    "    DNA = generate_random_sequence(length, dist)\n",
    "    \n",
    "    if DNA[k:k+6] == \"GAATTC\" :\n",
    "        print(DNA)\n",
    "        print(f\"Found 'ATG'. It took {j} turns \")\n",
    "        break\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e48356d7-b7b4-4ef9-a6ed-b965c4028194",
   "metadata": {},
   "source": [
    "**Part 5 - Introduction to Biopython**\n",
    "\n",
    "We will use the Biopython module for the next part.\n",
    "It does not come installed by default, so this will install it using pip.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa606af-0c4b-4ab3-b381-632478dadf9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28937a97-5572-4eb0-85d5-664971ee0f00",
   "metadata": {},
   "source": [
    "Biopython has a `Seq` function, which can generate a special string object that is specifically e.g. DNA or protein sequence:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c22044d1-018c-47fc-a5da-ac6b944d3b47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# @title Biopython\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "DNA = Seq(\"ATCGATGATAGGATA\")\n",
    "protein = Seq(\"CATINTHAMAT\")\n",
    "\n",
    "print(type(DNA))\n",
    "print(type(protein))\n",
    "\n",
    "#Usual string operations apply\n",
    "print(DNA[0:3])\n",
    "print(len(protein))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00bd2b28-eef1-47af-bc6f-ffe1797a9972",
   "metadata": {},
   "source": [
    "Biopython Seq type uses `methods`, which are similar to functions, but specific to a particular type of object. \n",
    "Examples:\n",
    "- make reverse complementary DNA sequence\n",
    "- transcibe or translate DNA to protein sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46ba571d-6b83-4d98-8546-8646cc93ed8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "reverse_complment = DNA.reverse_complement()\n",
    "transcribe = DNA.transcribe()\n",
    "translate = DNA.translate()\n",
    "\n",
    "\n",
    "print(DNA)\n",
    "print(reverse_complment)\n",
    "print(transcribe)\n",
    "print(translate)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "866291f4-e7fe-43ea-b2d1-8fc66461167e",
   "metadata": {},
   "source": [
    "Perhaps more usefully, Biopython can retrieve entries from the NCBI:\n",
    "\n",
    "https://biopython.org/docs/1.75/api/Bio.Entrez.html\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html\n",
    "\n",
    "\n",
    "\n",
    "Let's look at some entries for the enzyme COMT.\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/datasets/gene/1312/\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07f11afe-cfb8-4f50-89f4-713a90a0736f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "\n",
    "Entrez.email = \"XXX@manchester.ac.uk\"  # Always tell NCBI who you are\n",
    "\n",
    "handle = Entrez.efetch(db=\"nucleotide\", id=\"NM_000754.4\", rettype=\"gb\", retmode=\"text\")\n",
    "print(handle.read())\n",
    "handle.close()\n",
    "\n",
    "handle = Entrez.efetch(db=\"pdb\", id=\"NM_000754.4\", rettype=\"gb\", retmode=\"text\")\n",
    "print(handle.read())\n",
    "handle.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55474c2d-d297-4763-b609-0a005a2b8d3b",
   "metadata": {},
   "source": [
    "We can also use python to work with pdb files (protein structures).\n",
    "\n",
    "We will use another module call py3Dmol for this and will use !wget to directly download a pdb file from rcsb.org.\n",
    "\n",
    "https://www.rcsb.org/structure/6I3C\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9d00fb1-ffc2-4cad-bb00-7e4410089ef9",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install py3Dmol # Installing biopython using pip\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7237864-70a9-4b8c-aa27-59a2dca080b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "!wget http://files.rcsb.org/download/6I3C.pdb # downloading our pdb file\n",
    "!grep ATOM 6I3C.pdb > protein.pdb\n",
    "!grep HETATM 6I3C.pdb > ligand.pdb\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddec37e9-a5dc-4c3d-b944-7e92c87e2c92",
   "metadata": {},
   "outputs": [],
   "source": [
    "# and use py3Dmol to view the file.\n",
    "\n",
    "import py3Dmol\n",
    "v = py3Dmol.view()\n",
    "v.addModel(open('protein.pdb').read())\n",
    "v.setStyle({'cartoon':{}})\n",
    "v.addModel(open('ligand.pdb').read())\n",
    "v.setStyle({'model':1},{'stick':{'colorscheme':'greenCarbon'}})\n",
    "v.zoomTo({'model':1})\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7de6eada-93ed-4334-85ad-68aa54450c88",
   "metadata": {},
   "source": [
    "Now it's play time...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d61686c3-4739-4311-b3e9-c9a53f02e6cc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python [conda env:anaconda3]",
   "language": "python",
   "name": "conda-env-anaconda3-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
